### PR TITLE
Reduce cache validity using BID expiration (fixes #31)

### LINKED ISSUE #31: Cache Token server credentials for the validity of the BID
After thinking this through, it looks like we should cache TokenServer credentials for the validity of the Browser ID assertion only and for a given BID because otherwise we don't validate the BID when using the cache.

### PR DESCRIPTION
**Pull request by leplatrem**
- [x] Hack some code that looks ok
- [x] Start again with tests :)

### CODE REVIEW

**Comment by leplatrem** [comment]
@Natim r?

File: syncto/authentication.py [hunk 1]
```diff
@@ -73,3 +78,34 @@ def build_sync_client(request):
...
+    rem = len(value) % 4
+    if rem > 0:
+        value += b'=' * (4 - rem)
+    try:
+        return base64.urlsafe_b64decode(value)
+    except TypeError as e:
+        raise ValueError(str(e))
+
+
+def _extract_ttl(bid_assertion):
```
Comment (Natim) [inline syncto/authentication.py]: `_extract_bid_assertion_ttl`

File: syncto/tests/test_authentication.py [hunk 2]
```diff
@@ -107,3 +107,89 @@ def test_in_case_decryption_fails_it_should_raise(self):
                     self.request.registry.cache, 'get',
                     return_value=tempered_cache):
                 self.assertRaises(CryptoError, build_sync_client, self.request)
+
+    def test_uses_ttl_from_settings_if_assertion_has_no_expiration(self):
+        # This assertion comes from PyBrowserID tests cases.
+        # https://github.com/mozilla/PyBrowserID
+        assertion = """
+        eyJjZXJ0aWZpY2F0ZXMiOlsiZXlKaGJHY2lPaUpTVXpFeU9DSjkuZXlKcGMzTWlPaUppY
```
Comment (Natim) [inline syncto/tests/test_authentication.py:118]: @rfk It seems to me that this assertion doesn't have the same format as the one Firefox Account is using. Should we support this as well?

```
{"certificates":["eyJhbGciOiJSUzEyOCJ9.eyJpc3MiOiJicm93c2VyaWQub3JnIiwiZXhwIjoxMzIxOTQxOTgzMDUwLCJwdWJsaWMta2V5Ijp7ImFsZ29yaXRobSI6IlJTIiwibiI6IjgxNjkxNTA0NTk0NDUzNDU1ODAxNTkyNDkxMzk2NDE4NzE2MTU0NDM4NDYwODIzMzA1MTIyODcwMDQ4NzM1MDI5MDkzMzkyMDY3MDE1MjA0NDgzMjU1NzYwNDA5NzEyODc3MTk0OTUzMTAwMTQ2NDUzNzM2NjE4NTUzNDcxMjI1ODkyMzM1ODUxMjY5MDAwNDAyMTk2ODE0MDkiLCJlIjoiNjU1MzcifSwicHJpbmNpcGFsIjp7ImVtYWlsIjoicnlhbkByZmsuaWQuYXUifX0.k_hhKX0TBVu2_k3m_nD5NUb_NKp__O-651mBEIwKSYdiNzt0BoVFCDTEnxHPMbBV2Zz94X8--TcUrbtEt1euKWVv3-51T9LAfuzHH_zCBQrUnld1ZWJjA3_9dHPy30g4LIOXe2VZgtOSoks2dQ8d3ok9RQ2P0NDG0u00g7yFz18"],
"assertion":"eyJhbGciOiJSUzY0In0.eyJleHAiOjEzMjE5Mjk0NzU4MjksImF1ZCI6Imh0dHA6Ly9teWZhdm9yaXRlYmVlci5vcmcifQ.AhgKd4ys4KqgHbXqCRKxGvYnVaI9L-ohXHy4IPUX9m_24Mg_bU6hdH13S6qgC-oHpiKp_Lizp4FF9TrPc60SEw"}
```

```
{"alg":"RS128"}
{"iss":"browserid.org","exp":1321941983050,"public-key":{"algorithm":"RS","n":"8169150459445345580159249139641871615443846082330512287004873502909339206701520448325576040971287719495310014645373661855347122589233585126900040219681409","e":"65537"},"principal":{"email":"ryan@rfk.id.au"}}
```

```
{"alg":"RS64"}
{"exp":1321929475829,"aud":"http://myfavoritebeer.org"}
```

File: syncto/tests/test_authentication.py [hunk 3]
```diff
@@ -107,3 +107,89 @@ def test_in_case_decryption_fails_it_should_raise(self):
...
+                with mock.patch.object(self.request.registry.cache, 'set') \
+                        as mocked_set:
+                    build_sync_client(self.request)
+                    args, _ = tuple(mocked_set.call_args_list[-1])
+                    ttl = args[-1]
+                    self.assertEqual(int(ttl), 300)
+
+    def test_uses_ttl_from_assertion_if_smaller_than_settings(self):
+        assertion = """
+        eyJhbGciOiJSUzI1NiJ9.eyJwdWJsaWMta2V5Ijp7ImFsZ29yaXRobSI6IkRTIiwiZyI6I
```
Comment (Natim) [inline syncto/tests/test_authentication.py]: This ones contains:
```
{"alg":"RS256"}
{"public-key":{"algorithm":"DS","g":"67e586460f333269b79afbcd082f36994c14ccdc0ad95ca2cf1d816c25de492db126d90976fcf1f7ab12a814d110c13eb4bf529b3c7fa705baede72fb9e69bf3733990263d303b83f20c87848b7a83187ed3660e6418e8f8c00516ef90beff48023a5d7479e2050db3ae60aafd9dd4a28956cb5ee74f0044761d8175bd45639","q":"e075a81310701daf4a73a414aabf6cfe4cfb63e3","p":"9cb1df087dc249c12fda9adbf6a180d406fe508bc6277120ddd7fc4ab5118faf4f8f3c255faf0ce24b32595a0fcd6751746fbd7f33ceee11a9c650c7bd017595f0b9c18f7512289128c45643a072014c50cc5b103808d6ef5701abcd5803081b215204b992373dab1d5a27f3cb01c6f858dd8d777efce806dabb6b825c1b57e7","y":"962ed58194722cb0416a53a79d6ea38b1bd62b1c965a04bf3c3cd6047b5ff0d0afea2ba9f54b1d8193e17aaf5eaa775000b05ce46bcfe6037212d829aacadc71a105130499dc69b3310e2dc21802a435abb543b091fce555fadb4d2dcbf5c7c16d99547d05c0d00d8d0bf92e0163e154bc442fa54c8f07c125834ed2ad226abb"},"principal":{"email":"46655d10425549639aae297c6fc6d632@api.accounts.firefox.com"},"iat":1444122573580,"exp":1444124383580,"fxa-generation":1444122508836,"fxa-lastAuthAt":1444122582,"fxa-verifiedEmail":"syncto@restmail.net","iss":"api.accounts.firefox.com"}
{"aud": "https://token.services.mozilla.com/", "exp": 1444126183000}
```

File: syncto/tests/test_authentication.py (hunk 2 above)
Comment (rfk) [inline syncto/tests/test_authentication.py:118]: These are "old-format" browserid assertions, persona.org used to produce things in that format but it hasn't for a long time and IIUC nothing else uses them.  So it should be safe to ignore it.

**Comment by Natim** [comment]
r+

**Pull request merged**